File: wassupweb/wabinary/jid_utils.py
from __future__ import annotations

from enum import IntEnum
from typing import TypedDict

S_WHATSAPP_NET = "s.whatsapp.net"
OFFICIAL_BIZ_JID = "16505361212@c.us"
SERVER_JID = "server@c.us"
PSA_WID = "0@c.us"
STORIES_JID = "status@broadcast"
META_AI_JID = "13135550002@c.us"


class WAJIDDomains(IntEnum):
    WHATSAPP = 0
    LID = 1
    HOSTED = 128
    HOSTED_LID = 129


class FullJid(TypedDict, total=False):
    user: str
    server: str
    device: int
    domainType: int


def get_server_from_domain_type(initial_server: str, domain_type: WAJIDDomains | None = None) -> str:
    if domain_type == WAJIDDomains.LID:
        return "lid"
    if domain_type == WAJIDDomains.HOSTED:
        return "hosted"
    if domain_type == WAJIDDomains.HOSTED_LID:
        return "hosted.lid"
    return initial_server


def jid_encode(user: str | int | None, server: str, device: int | None = None, agent: int | None = None) -> str:
    user_part = "" if user is None else str(user)
    agent_part = f"_{agent}" if agent else ""
    device_part = f":{device}" if device else ""
    return f"{user_part}{agent_part}{device_part}@{server}"


def jid_decode(jid: str | None) -> FullJid | None:
    if not isinstance(jid, str):
        return None
    sep_idx = jid.find("@")
    if sep_idx < 0:
        return None

    server = jid[sep_idx + 1 :]
    user_combined = jid[:sep_idx]
    user_agent, *device_part = user_combined.split(":")
    user, *agent_part = user_agent.split("_")
    agent = agent_part[0] if agent_part else None
    device = int(device_part[0]) if device_part and device_part[0] else None

    domain_type = WAJIDDomains.WHATSAPP
    if server == "lid":
        domain_type = WAJIDDomains.LID
    elif server == "hosted":
        domain_type = WAJIDDomains.HOSTED
    elif server == "hosted.lid":
        domain_type = WAJIDDomains.HOSTED_LID
    elif agent:
        domain_type = WAJIDDomains(int(agent))

    return FullJid(server=server, user=user, domainType=int(domain_type), device=device)


def are_jids_same_user(jid1: str | None, jid2: str | None) -> bool:
    left = jid_decode(jid1)
    right = jid_decode(jid2)
    return bool(left and right and left.get("user") == right.get("user"))


def is_jid_meta_ai(jid: str | None) -> bool:
    return bool(jid and jid.endswith("@bot"))


def is_pn_user(jid: str | None) -> bool:
    return bool(jid and jid.endswith("@s.whatsapp.net"))


def is_lid_user(jid: str | None) -> bool:
    return bool(jid and jid.endswith("@lid"))


def is_jid_broadcast(jid: str | None) -> bool:
    return bool(jid and jid.endswith("@broadcast"))


def is_jid_group(jid: str | None) -> bool:
    return bool(jid and jid.endswith("@g.us"))


def is_jid_status_broadcast(jid: str) -> bool:
    return jid == "status@broadcast"


def is_jid_newsletter(jid: str | None) -> bool:
    return bool(jid and jid.endswith("@newsletter"))


def is_hosted_pn_user(jid: str | None) -> bool:
    return bool(jid and jid.endswith("@hosted"))


def is_hosted_lid_user(jid: str | None) -> bool:
    return bool(jid and jid.endswith("@hosted.lid"))


def is_jid_bot(jid: str | None) -> bool:
    if not jid:
        return False
    user = jid.split("@", 1)[0]
    if not jid.endswith("@c.us"):
        return False
    return (len(user) == 11 and user.startswith("1313555")) or (len(user) == 11 and user.startswith("131655500"))


def jid_normalized_user(jid: str | None) -> str:
    result = jid_decode(jid)
    if not result:
        return ""
    user = result["user"]
    server = result["server"]
    return jid_encode(user, "s.whatsapp.net" if server == "c.us" else server)


def transfer_device(from_jid: str, to_jid: str) -> str:
    from_decoded = jid_decode(from_jid) or {}
    to_decoded = jid_decode(to_jid)
    if not to_decoded:
        raise ValueError(f"invalid JID: {to_jid}")
    device_id = from_decoded.get("device", 0) or 0
    return jid_encode(to_decoded["user"], to_decoded["server"], device_id)

File: wassupweb/src/wabinary/jid_utils.rs

#[derive(Debug, Clone, Copy, PartialEq, Eq)]
pub enum WAJIDDomains {
    Whatsapp = 0,
    Lid = 1,
    Hosted = 128,
    HostedLid = 129,
}

impl WAJIDDomains {
    pub fn from_u32(value: u32) -> Option<Self> {
        match value {
            0 => Some(WAJIDDomains::Whatsapp),
            1 => Some(WAJIDDomains::Lid),
            128 => Some(WAJIDDomains::Hosted),
            129 => Some(WAJIDDomains::HostedLid),
            _ => None,
        }
    }
}

#[derive(Debug, Clone, PartialEq, Eq)]
pub struct FullJid {
    pub user: String,
    pub server: String,
    pub device: Option<u32>,
    pub domain_type: u32,
}

pub fn get_server_from_domain_type(initial_server: &str, domain_type: Option<WAJIDDomains>) -> &str {
    match domain_type {
        Some(WAJIDDomains::Lid) => "lid",
        Some(WAJIDDomains::Hosted) => "hosted",
        Some(WAJIDDomains::HostedLid) => "hosted.lid",
        _ => initial_server,
    }
}

pub fn jid_encode(
    user: Option<&str>,
    server: &str,
    device: Option<u32>,
    agent: Option<u32>,
) -> String {
    let mut result = String::new();
    if let Some(u) = user {
        result.push_str(u);
    }
    if let Some(a) = agent {
        result.push_str(&format!("_{}", a));
    }
    if let Some(d) = device {
        result.push_str(&format!(":{}", d));
    }
    result.push('@');
    result.push_str(server);
    result
}

pub fn jid_decode(jid: Option<&str>) -> Option<FullJid> {
    let jid = jid?;
    let sep_idx = jid.find('@')?;
    let (user_combined, server) = jid.split_at(sep_idx);
    let server = &server[1..]; // skip '@'

    let mut parts = user_combined.split(':');
    let user_agent = parts.next().unwrap_or("");
    let device_str = parts.next();
    let device = device_str.and_then(|s| s.parse::<u32>().ok());

    let mut user_parts = user_agent.split('_');
    let user = user_parts.next().unwrap_or("").to_string();
    let agent_str = user_parts.next();
    let agent = agent_str.and_then(|s| s.parse::<u32>().ok());

    let mut domain_type = WAJIDDomains::Whatsapp as u32;
    if server == "lid" {
        domain_type = WAJIDDomains::Lid as u32;
    } else if server == "hosted" {
        domain_type = WAJIDDomains::Hosted as u32;
    } else if server == "hosted.lid" {
        domain_type = WAJIDDomains::HostedLid as u32;
    } else if let Some(a) = agent {
        domain_type = a;
    }

    Some(FullJid {
        user,
        server: server.to_string(),
        device,
        domain_type,
    })
}

pub fn are_jids_same_user(jid1: Option<&str>, jid2: Option<&str>) -> bool {
    let j1 = jid_decode(jid1);
    let j2 = jid_decode(jid2);
    match (j1, j2) {
        (Some(j1), Some(j2)) => j1.user == j2.user,
        _ => false,
    }
}

pub fn is_jid_meta_ai(jid: Option<&str>) -> bool {
    jid.map_or(false, |s| s.ends_with("@bot"))
}

pub fn is_pn_user(jid: Option<&str>) -> bool {
    jid.map_or(false, |s| s.ends_with("@s.whatsapp.net"))
}

pub fn is_lid_user(jid: Option<&str>) -> bool {
    jid.map_or(false, |s| s.ends_with("@lid"))
}

pub fn is_jid_broadcast(jid: Option<&str>) -> bool {
    jid.map_or(false, |s| s.ends_with("@broadcast"))
}

pub fn is_jid_group(jid: Option<&str>) -> bool {
    jid.map_or(false, |s| s.ends_with("@g.us"))
}

pub fn is_jid_status_broadcast(jid: &str) -> bool {
    jid == "status@broadcast"
}

pub fn is_jid_newsletter(jid: Option<&str>) -> bool {
    jid.map_or(false, |s| s.ends_with("@newsletter"))
}

pub fn is_hosted_pn_user(jid: Option<&str>) -> bool {
    jid.map_or(false, |s| s.ends_with("@hosted"))
}

pub fn is_hosted_lid_user(jid: Option<&str>) -> bool {
    jid.map_or(false, |s| s.ends_with("@hosted.lid"))
}

pub fn is_jid_bot(jid: Option<&str>) -> bool {
    if let Some(j) = jid {
        if !j.ends_with("@c.us") {
            return false;
        }
        if let Some((user, _)) = j.split_once('@') {
            return (user.len() == 11 && user.starts_with("1313555")) || (user.len() == 11 && user.starts_with("131655500"));
        }
    }
    false
}

pub fn jid_normalized_user(jid: Option<&str>) -> String {
    if let Some(result) = jid_decode(jid) {
        let server = if result.server == "c.us" {
            "s.whatsapp.net"
        } else {
            &result.server
        };
        jid_encode(Some(&result.user), server, None, None)
    } else {
        String::new()
    }
}

pub fn transfer_device(from_jid: &str, to_jid: &str) -> Result<String, String> {
    let from_decoded = jid_decode(Some(from_jid));
    let to_decoded = jid_decode(Some(to_jid)).ok_or_else(|| format!("invalid JID: {}", to_jid))?;

    let device_id = from_decoded.and_then(|d| d.device).unwrap_or(0);
    Ok(jid_encode(Some(&to_decoded.user), &to_decoded.server, (device_id > 0).then_some(device_id), None))
}

#[cfg(test)]
mod tests {
    use super::*;

    #[test]
    fn test_jid_encode() {
        assert_eq!(
            jid_encode(Some("123"), "s.whatsapp.net", None, None),
            "123@s.whatsapp.net"
        );
        assert_eq!(
            jid_encode(Some("123"), "s.whatsapp.net", Some(1), None),
            "123:1@s.whatsapp.net"
        );
        assert_eq!(
            jid_encode(Some("123"), "s.whatsapp.net", Some(1), Some(2)),
            "123_2:1@s.whatsapp.net"
        );
        assert_eq!(
            jid_encode(None, "server@c.us", None, None),
            "@server@c.us"
        );
    }

    #[test]
    fn test_jid_decode() {
        let jid = jid_decode(Some("123@s.whatsapp.net")).unwrap();
        assert_eq!(jid.user, "123");
        assert_eq!(jid.server, "s.whatsapp.net");
        assert_eq!(jid.device, None);
        assert_eq!(jid.domain_type, WAJIDDomains::Whatsapp as u32);

        let jid2 = jid_decode(Some("123:1@s.whatsapp.net")).unwrap();
        assert_eq!(jid2.user, "123");
        assert_eq!(jid2.server, "s.whatsapp.net");
        assert_eq!(jid2.device, Some(1));

        let jid3 = jid_decode(Some("123_2:1@s.whatsapp.net")).unwrap();
        assert_eq!(jid3.user, "123");
        assert_eq!(jid3.server, "s.whatsapp.net");
        assert_eq!(jid3.device, Some(1));
        assert_eq!(jid3.domain_type, 2);

        let jid_lid = jid_decode(Some("123@lid")).unwrap();
        assert_eq!(jid_lid.domain_type, WAJIDDomains::Lid as u32);

        let jid_hosted = jid_decode(Some("123@hosted")).unwrap();
        assert_eq!(jid_hosted.domain_type, WAJIDDomains::Hosted as u32);

        assert!(jid_decode(Some("invalid")).is_none());
        assert!(jid_decode(None).is_none());
    }

    #[test]
    fn test_is_functions() {
        assert!(is_pn_user(Some("123@s.whatsapp.net")));
        assert!(!is_pn_user(Some("123@g.us")));

        assert!(is_lid_user(Some("123@lid")));
        assert!(!is_lid_user(Some("123@s.whatsapp.net")));

        assert!(is_jid_group(Some("123@g.us")));

        assert!(is_jid_bot(Some("13135551234@c.us")));
        assert!(!is_jid_bot(Some("123@c.us")));
    }

    #[test]
    fn test_are_jids_same_user() {
        assert!(are_jids_same_user(Some("123:1@s.whatsapp.net"), Some("123:2@s.whatsapp.net")));
        assert!(!are_jids_same_user(Some("123@s.whatsapp.net"), Some("456@s.whatsapp.net")));
        assert!(!are_jids_same_user(None, Some("123@s.whatsapp.net")));
    }

    #[test]
    fn test_jid_normalized_user() {
        assert_eq!(jid_normalized_user(Some("123:1@c.us")), "123@s.whatsapp.net");
        assert_eq!(jid_normalized_user(Some("123@lid")), "123@lid");
        assert_eq!(jid_normalized_user(None), "");
    }

    #[test]
    fn test_transfer_device() {
        assert_eq!(transfer_device("123:1@s.whatsapp.net", "456@s.whatsapp.net").unwrap(), "456:1@s.whatsapp.net");
        assert_eq!(transfer_device("123@s.whatsapp.net", "456:2@s.whatsapp.net").unwrap(), "456@s.whatsapp.net");
    }
}
